### ISSUE TITLE
`ScalableImageWidget` is flashing between `Route`s even if the `cache` is set

### ISSUE THREAD

**Issue by AlexV525** [zathras/jovial_svg#48]
## Describe the bug

When I use the `ScalableImageWidget` wrapped by `Hero` between different `Route`s with the `cache` set to `ScalableImageCache`, images flash during route changing.

## Additional context

I've updated the code based on `example/lib/cache.dart` at https://github.com/AlexV525/jovial_svg/commit/90e1b2e88bb0c4977bbb79723be14882a03c77d2.

Few conditions I've considered:
- They're using the same top cache. So it shouldn't relate to mapping issues.
- I've noticed that `_CacheEntry._si` is `Future<ScalableImage>`, which seems like it needs to wait again with every read call.

Screen record based on the repro code:

https://user-images.githubusercontent.com/15884415/218276551-710840d9-fd8c-42e3-bbd4-b8f8c2881f88.mp4

## Questions (thoughts)

- What if we change `_CacheEntry._si` to `ScalableImage` (if it's the root cause)? Did you mean to set it as Future because storing the image could consume uncontrollable memories?
- Can we use a caching mechanism like the `PaintingBinding.instance.imageCache`?

**Comment by AlexV525** [comment]
I've done some investigation into this, in order to make it fit the request, we need to remove the `Future` from `_si`, and try to fetch the cache entry once the `_AsyncSIWidget` started to init. See https://github.com/AlexV525/jovial_svg/commit/c8d9c53f4a4daa5439a100cd3a6d0dda4ba3896b.

Would like to consult your opinions. The unawaited `addReference` might have some race-condition issues here.

**Comment by zathras** [comment]
What's `Hero`?  Where is full, buildable source?  Impossible to reproduce without complete information.

However, I strongly suspect it's a bug in one of the parent widgets, possibly `Hero`.  Or possibly you need to manage the `Key` for more of the stateful widgets.

About it being a `Future`, no, it's intentionally a future.  Think about it:  When there's a cache miss, it needs to fetch the image asynchronously.

If it's difficult to fix `Hero` (assuming that be where the problem lies) or manage keys or whatever the problem is, you could always just maintain your own cache of `ScalableImage` instances, at the level of your stateful widget, and then use the ScalableImageWidget constructor that takes an instance.  Dart's model for maintaining state when the screen is re-built can get tricky, at times.

**Comment by AlexV525** [comment]
> What's `Hero`? Where is full, buildable source? Impossible to reproduce without complete information.

As mentioned in https://github.com/zathras/jovial_svg/issues/48#issue-1580979687: https://github.com/AlexV525/jovial_svg/tree/90e1b2e88bb0c4977bbb79723be14882a03c77d2

**Comment by AlexV525** [comment]
> If it's difficult to fix `Hero` (assuming that be where the problem lies) or manage keys or whatever the problem is, you could always just maintain your own cache of `ScalableImage` instances, at the level of your stateful widget, and then use the ScalableImageWidget constructor that takes an instance. Dart's model for maintaining state when the screen is re-built can get tricky, at times.

Building custom caches is easy, but building a custom `ScalableImageWidget` with a custom cache is not easy in this case. Instead, updating caches will require building a new widget other than `ScalableImageWidget`.

**Comment by zathras** [comment]
And what happens when you provide an appropriate `Key`?

**Comment by AlexV525** [comment]
> And what happens when you provide an appropriate `Key`?

It just doesn't help, unfortunately. :(

**Comment by zathras** [comment]
This is interesting.  It looks like Flutter's rendering is less robust than had been apparent, when an already-completed Future is involved in building a UI.  I'm not sure if this is something to do with `Hero` being more aggressive than other widgets, or what exactly.

Annoyingly, `Future` doesn't have a way of saying "if this future has already completed, just give me the value synchronously," so I had to work around that with a "give me the value if known synchronously" parameter to the relevant cache method.

Oh, and BTW, when I first looked at your repo, I could see the diffs of your change, but the committed version wasn't showing up.  Maybe some kind of caching problem at github?  Or maybe I clicked a wrong link or something.  Anyway, it's there now.  Generally, it's simpler to just attach code used to reproduce an issue to the issue itself.

Anyway, try with https://github.com/zathras/jovial_svg/commit/f890f9272ae5d38e576f661c52412c4947c4535b

**Comment by AlexV525** [comment]
> Anyway, try with [f890f92](https://github.com/zathras/jovial_svg/commit/f890f9272ae5d38e576f661c52412c4947c4535b)

Thanks, I've checked with the commit and it seems working. Closing the issue now.

> This is interesting. It looks like Flutter's rendering is less robust than had been apparent, when an already-completed Future is involved in building a UI. I'm not sure if this is something to do with `Hero` being more aggressive than other widgets, or what exactly.

What Flutter does is diff the image state based on the image key, and expose an argument named `gaplessPlayback` that pulls out the old image.

> Annoyingly, `Future` doesn't have a way of saying "if this future has already completed, just give me the value synchronously," so I had to work around that with a "give me the value if known synchronously" parameter to the relevant cache method.

Rather than using the `Future`, you may try `Completer` which can act like locks. Some pseudo codes:
```dart
class _CacheEntry {
  ScalableImage? _siSync;
  Completer<ScalableImage>? _completer;

  Future<ScalableImage> addReference(...) {
    _CacheEntry? e = _canonicalized[src];
    if (e._siSync != null) { return e._siSync!; }
    if (_completer == null) {
      _completer = Completer<ScalableImage>();
      e._si.then((ScalableImage r) {
        _completer.complete(r);
      }).catchError((Object e, StackTrace s) {
        _completer.completeError(e, s);
      }).whenComplete(() {
        _completer = null;
      });
    }
    return _completer!.future;
  }
}
```

**Comment by zathras** [comment]
Sure, other implementations are possible.

Re-opening issue:  I don't close issues in this repo until I've published the fix.

**Comment by zathras** [comment]
Published in 1.1.12
